Fix Issue 23709 - Cannot use synchronized on shared class with -preview=nosharedaccess

RazvanN7 opened the pull request.

WalterBright committed:
fix Issue 23387 - ImportC: identical structs defined in two C files lead to duplicate .init symbol on macOS

dkorpel committed:
Implement named args resolution

JohanEngelen committed:
Add -verror-supplements, to let the user define how many supplemental error messages he wants to see.

In particular this prevents having to rebuild with "-v" flag for ... (18 instantiations, -v to show) to show (and with all the other -v noise).

WalterBright committed:
fix Issue 19575 - core.cpuid not usable without a runtime

WalterBright committed:
replace addLocalClass() with ScopeDsymbol._foreach

WalterBright committed:
fix Issue 21772 - [REG2.069] Consecutive different-signed double.nans in an array literal take the sign of the previous nan (same for float and real)

ibuclaw committed:
Update copyright years of dmd sources to 2023

ibuclaw committed:
Merge remote-tracking branch 'upstream/stable' into merge_stable

Commit:
Merge pull request #14767 from ibuclaw/merge_stable

bump VERSION to v2.102.0-beta.1

Signed-off-by: Petar Kirov <PetarKirov@users.noreply.github.com>
Merged-on-behalf-of: Petar Kirov <PetarKirov@users.noreply.github.com>

WalterBright committed:
cdmemset: remove code replaced by elmemset (#14769)

dkorpel committed:
Remove old 'Calling member function before dip1000'

ibuclaw committed:
Makefile: stub out all auto-tester-build and auto-tester-test recipes

ibuclaw committed:
ci: Update cirrus macOS image to M1-based runners (#14772)

* druntime/test: Force makefile to x86 on Apple ARM cpu (dmd does not support ARM)

* ci: Update cirrus macOS image to M1-based runners

* compiler/test/Makefile: Workaround issue 23517

WalterBright committed:
handle byte, short, and int cases for bug 19946 (#14770)

ntrel committed:
[dsymbolsem] Tweak punctuation for 3 error messages (#14780)

* Tweak punctuation for 3 error messages

Fix Issue 13656 - clarify error when trying to declare a variable of
type ref.
Tweak 2 inout errors.

* Use hyphen & fix `auto Type` error too

ntrel committed:
[dsymbolsem] Fix error punctuation (part 2) (#14784)

ntrel committed:
[dsymbolsem] Fix error punctuation (part 3) (#14785)

ntrel committed:
[dsymbolsem] Fix error punctuation (part 4) (#14786)

* [dsymbolsem] Fix error punctuation (part 4)

Also move fail233.d test into diag12312.d.

* 3 pragma(mangle) errors

* Fix line no

CyberShadow committed:
Fix Issue 23597 - .di files not compatible with -i (#14783)

WalterBright committed:
add support for 64 bit memset (#14788)

RazvanN7 committed:
Fix Issue 23261 - druntime core.std.attribute.Tagged1_2 constructor is unsafe (#14771)

WalterBright committed:
fix Issue 20101 - BetterC: Template instantiation in CTFE only context should skip codegen (#14789)

CyberShadow committed:
core.sys.posix.sys.stat: Fix duplicate definition of __mode_t on Musl

The definition was conflicting with the other definitions within the
same file.

WalterBright committed:
fix Issue 23584 - ImportC: __builtin_bit_cast not supported (#14795)

WalterBright committed:
ImportC better error message for __int128 (#14797)

WalterBright committed:
fix Issue 23606 - betterC with CTFE and gc (#14791)

WalterBright committed:
fix Issue 23616 - ImportC: clang __has_feature and __has_extension not recognized (#14799)

WalterBright committed:
fix Issue 23583 - ImportC: undefined identifier __builtin___memmove_chk

dkorpel committed:
Follow up "- betterC with CTFE and gc" (#14800)

RazvanN7 committed:
Fix Issue 23617 - traits(child) compile error need this for something that doesn't need this

WalterBright committed:
fix Issue 23407 - ImportC: function-local struct definition as part of variable declaration doesn.t shadow global definition (#14806)

dkorpel committed:
Handle `null` elements in `Array.toString`

WalterBright committed:
fix Issue 23622 - ImportC #defines conflict with declarations

Commit:
Merge pull request #14813 from dlang/stable

Merge Stable into master

Signed-off-by: Nicholas Wilson <thewilsonator@users.noreply.github.com>
Merged-on-behalf-of: Dennis <dkorpel@users.noreply.github.com>

KytoDragon committed:
Fix 23625 - Add missing memory functions for windows

Commit:
Merge pull request #14815 from dlang/stable

Merge stable

Signed-off-by: Nicholas Wilson <thewilsonator@users.noreply.github.com>
Merged-on-behalf-of: Dennis <dkorpel@users.noreply.github.com>

ntrel committed:
Fix Issue 23624 - Race condition in test runner for DMD

ibuclaw committed:
follow-up Issue 23626 - invert trustSystemEqualsDefault initializer to true

This is the desired behaviour for any future users of this function.

ntrel committed:
Declare failedTargets shared

Commit:
Merge pull request #14817 from ntrel/dmd-test-fails

Fix Issue 23624 - Race condition in test runner for DMD

Signed-off-by: Dennis <dkorpel@users.noreply.github.com>
Merged-on-behalf-of: Iain Buclaw <ibuclaw@users.noreply.github.com>

ibuclaw committed:
Merge remote-tracking branch 'upstream/stable' into merge_stable

ibuclaw committed:
Merge pull request #14823 from ibuclaw/merge_stable

merge stable

ibuclaw committed:
ci: Update Ubuntu images to 20.04

ibuclaw committed:
ci: Update CirrusCI Ubuntu images to 22.04

ibuclaw committed:
ci: Remove old druntime CI scripts

ibuclaw committed:
runnable_cxx: Remove new_allocator test from cppa.d

ibuclaw committed:
importc: Remove __has_builtin redefine

ibuclaw committed:
ci: Workaround gdb bug in by requesting privileged container

ibuclaw committed:
src/dmd/errors.d(384): `old_loc` is thread local

ljmf00 committed:
druntime: refactor PAGESIZE to use core.memory.pageSize

Signed-off-by: Luís Ferreira <contact@lsferreira.net>

ljmf00 committed:
druntime: avoid using compiler tricks of unspeciefied .mangleof (#14832)

Signed-off-by: Luís Ferreira <contact@lsferreira.net>

Signed-off-by: Luís Ferreira <contact@lsferreira.net>

WalterBright committed:
fix Issue 21492 - betterC: TypeInfo is generated for code guarded by if(__ctfe) (#14830)

dkorpel committed:
Add named arguments to struct literals (#14776)

* Add named argument parsing

* Refactor StructInitializer semantic to make it reusable

* Move out `resolveStructLiteralNamedArgs`

* Resolve named args in struct literal

* Add tests

* Check for errors returned by `resolveStructLiteralNamedArgs`

* Expand names along with tuples

* Update error messages in tests

* Convert tabs to spaces in named arg test

* Fix style of `resolveStructLiteralNamedArgs`

* Clarify 'fixme' comment in expandTuples

* Add example to `expandTuples` documentation

* Update compiler/src/dmd/initsem.d

Co-authored-by: Razvan Nitu <razvan.nitu1305@gmail.com>

* Add overlapping initialization supplemental error to Struct Initializer

* Improve "too many initializers" error message

Co-authored-by: Razvan Nitu <razvan.nitu1305@gmail.com>

WalterBright committed:
remove untag.d from test suite (#14833)

dkorpel committed:
Fix parens in expandTuples documentation

CyberShadow committed:
dmd.expression: Fix another stray paren in DDoc

CyberShadow committed:
Merge pull request #14835 from CyberShadow/pull-20230118-030238

dmd.expression: Fix another stray paren in DDoc

dkorpel committed:
Pass names along with arguments

RazvanN7 committed:
Deprecate traits(isVirtualFunction) and traits(getVirtualFunctions) (#14802)

RazvanN7 committed:
Fix Issue 23639 - Casting to shared not allowed with -preview=nosharedaccess (#14836)

dkorpel committed:
Don't reorder arguments too soon

ntrel committed:
Fix Issue 23635 - Nonsensical "`case` must be a `string` or an integral constant, not `x`"

WalterBright committed:
fix Issue 16495 - __traits(fullyQualifedName) instead of std.traits.fullyQualifiedName (#14711)

WalterBright committed:
AliasDeclaration semantic

WalterBright committed:
fix Issue 23651 - Order dependency in semantic analysis of template members

ntrel committed:
Fix issue 11051 - Unmatched case in a final switch should throw in both release and non-release mode (#14841)

* Fix issue 11051

Keep a HALT instruction in a final switch statement if
the function is @safe and -release mode is enabled.

* @system -release test should also be non-zero

* don't suppress test errors

* Remove unreported error line

* Disable core dumps

Co-authored-by: Andrej Mitrovic <andrej.mitrovich@gmail.com>

Commit:
Merge pull request #14843 from WalterBright/aliassem

fix Issue 23651 - Order dependency in semantic analysis of template members

Merged-on-behalf-of: Iain Buclaw <ibuclaw@users.noreply.github.com>

WalterBright committed:
rename VarDeclaration.aliassym to aliasTuple (#14845)

RazvanN7 committed:
Propagate SCOPE.compiles for nested functions in traits(compiles) blocks (#14846)

RazvanN7 committed:
Remove FUNCFLAG.compileTimeOnly

dkorpel committed:
Add named arguments to `new class`

WalterBright committed:
core.int128 alignment is adjusted (#14768)

WalterBright committed:
fix Issue 23598 - Circular reference bug with static if and eponymous templates (#14838)

dkorpel committed:
Add test for calling a function type (#14848)

RazvanN7 committed:
Fix Issue 23650 - Using typeid with struct defined in __traits(compiles, ...) causes linker error (#14844)

dkorpel committed:
Improve `resolveNamedArgs` signature

dkorpel committed:
Implement named arguments for `new` expressions

teodutu committed:
Translate `_d_newclass` to a template (#14837)

This makes the following changes:
- Add template `_d_newclassT` to `druntime.src.core.lifetime.d`
- Replace lowering of `new C()` to `_d_newclassT!C()`
- Add `lowering` member to `NewExp`. This field stores the above
lowering to be used by e2ir.d
- Keep the old `_d_newclass` hook because it's used by
`TypeInfo_Class.create()`
- Add dummy `_d_newclassT` hook to tests that redefine the `object`
module
- Remove `new MinHeap!(TestType)()` from `fail308.d`. Otherwise the
errror was changed and printed the local path to druntime
- Move `err` to global scope in rt.sections.d to avoid the frontend
lowering
- Account for the `GC.malloc()` called by the template hook in the
`-profile=gc` tests

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

ntrel committed:
Tweak @live undisposed owner error

A pointer 'left dangling' means it refers to an invalid memory address -
change to 'not disposed of'. This also is consistent with the 'assigning
to Owner without disposing of owned value' error message.

ibuclaw committed:
ci: Version out C++ test that doesn't work with DMD

ibuclaw committed:
druntime: Match either _Dmain or D main

Commit:
Merge pull request #14824 from ibuclaw/ci-ubuntu-image

ci: Update Ubuntu images to 22.04

Signed-off-by: Nicholas Wilson <thewilsonator@users.noreply.github.com>
Merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

WalterBright committed:
fix Issue 23658 - .di generation of variables should turn them into declarations (#14851)

ibuclaw committed:
Merge remote-tracking branch 'upstream/stable' into merge_stable

Commit:
Merge pull request #14857 from ibuclaw/merge_stable

merge stable

Signed-off-by: Nicholas Wilson <thewilsonator@users.noreply.github.com>
Merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

benjones committed:
add const scope to IN6* functions taking inet 6 address pointers

RazvanN7 committed:
Fix Issue 23648 - replace uses of sprintf with snprintf in the compiler (#14854)

dkorpel committed:
Fix named arguments with `opCall`

dkorpel committed:
print named arguments in error messages

dkorpel committed:
Handle named arguments in template deduction

dkorpel committed:
Remove dead `return ref` check

ibuclaw committed:
dmd.root.filename: Add FileName::create helper

ibuclaw committed:
reapply fix Issue 23479 - ImportC recognizes .i and .c files, but not .h files

dkorpel committed:
d_do_test: allow separately compiling c sources

WalterBright committed:
fix Issue 20781 - Can call @live function without checking dip1021 rules (#11071)

WalterBright committed:
ImportC: add __check(assign-expression) extension (#14026)

Commit:
Merge pull request #14541 from WalterBright/fix23387

fix Issue 23387 - ImportC: identical structs defined in two C files l…

Signed-off-by: Dennis <dkorpel@users.noreply.github.com>
Merged-on-behalf-of: Dennis <dkorpel@users.noreply.github.com>

WalterBright committed:
fix Issue 21821 - Optimizer assumes immutables do not change, but they can in @system code (#12424)

WalterBright committed:
add scope to constructors (#14323)

WalterBright committed:
fix Issue 15985 - [REG2.068/2.069] Code doesn't link unless compiled with -debug (#14855)

WalterBright committed:
fix Issue 23545 - export int a; should generate dllexport, not dllimport (#14680)

WalterBright committed:
fix Issue 23145 - Stack allocation of scope new variables defeats @safe (#14175)

WalterBright committed:
fix Issue 23195 - Win64 function ABI bug for small non-POD arguments (#14651)

WalterBright committed:
fix Issue 23558 - add __traits(getModuleClasses [, module name]) (#14699)

ibuclaw committed:
fix Issue 23688 - FTBFS: error: cannot convert 'Expression' to 'Expression*'

dkorpel committed:
Revert "deprecate Object.factory"

This reverts commit 5600cfb42f34e318289102ddeba8b2fdb75deb97.

WalterBright committed:
fix Issue 23682 - dip1000 problem with return by ref (#14869)

ibuclaw committed:
Merge remote-tracking branch 'upstream/stable' into merge_stable

Commit:
Merge pull request #14875 from ibuclaw/merge_stable

merge stable

Signed-off-by: Nicholas Wilson <thewilsonator@users.noreply.github.com>
Merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

dkorpel committed:
Address Walter's review

dkorpel committed:
Add test for unknown named argument

dlang-bot committed:
Merge pull request #14575 from dkorpel/named-arg-resolve

Implement named argument for functions

Signed-off-by: Razvan Nitu <razvan.nitu1305@gmail.com>
Signed-off-by: Walter Bright <WalterBright@users.noreply.github.com>
Merged-on-behalf-of: unknown

WalterBright committed:
Revert "reapply fix Issue 23479 - ImportC recognizes .i and .c files, but not .h files" (#14870)

This reverts commit 650b75b7c19a3e9fd8c8645ac538f8e08bedc733.

WalterBright committed:
gitignore: ignore vim backup files

WalterBright committed:
reboot separate out mangleType() (#14862)

dkorpel committed:
Remove old dead code from parse.d

dkorpel committed:
Fix inconsistent names and arguments arrays

ibuclaw committed:
[refactor] Expression: change parens type from ubyte to bool

dkorpel committed:
Allow default arguments in the middle (#14879)

RazvanN7 committed:
Deprecate alias this for classes v2 (#14812)

RazvanN7 committed:
Fix Issue 23732 - Cannot create shared instance of class with -preview=nosharedaccess

RazvanN7 commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -13174,10 +13174,6 @@ bool checkSharedAccess(Expression e, Scope* sc, bool returnRef = false)
-            // Note: This handles things like `new shared(Throwable).msg`,
-            // where accessing `msg` would violate `shared`.
-            if (e.newtype.isShared())
-                return sharedError(original);
```
This should be checked for in DotVarExp, not here since it essentially forbids creating any shared instances via new.

dkorpel commented on compiler/test/fail_compilation/shared.d:
```diff
@@ -87,7 +87,6 @@ fail_compilation/shared.d(2154): Error: direct access to shared `*c.c1` is not a
 fail_compilation/shared.d(2160): Error: direct access to shared `*c.c1.c1` is not allowed, see `core.atomic`
 fail_compilation/shared.d(2181): Error: direct access to shared `k` is not allowed, see `core.atomic`
 fail_compilation/shared.d(2187): Error: direct access to shared `k.k2.k1` is not allowed, see `core.atomic`
-fail_compilation/shared.d(2194): Error: direct access to shared `(new shared(K2)).k1` is not allowed, see `core.atomic`
```
Because your code example returns a compile-time known pointer to `c`, while the code in the test case would return the shared pointer value of `K2` + the offset of `k1`.

dkorpel replied:
I think it could be allowed on the grounds that `new shared (K2)` is isolated (no aliases).

RazvanN7 replied:
Ok, are there any changes that you require for this PR?

dkorpel replied:
Someone wrote code and a test explicitly for `(new shared(K2)).k1`, this bug fix should not introduce new semantics to `shared` without approval.

dkorpel replied:
Also, as noted in bugzilla there's two issues:
- cannot create shared class instance
- cannot use `synchronized`
This PR should either fix both, or rename the issue to the one that it fixes.

ibuclaw committed:
Merge remote-tracking branch 'upstream/stable' into merge_stable

dlang-bot committed:
Merge pull request #14887 from dlang/merge_stable

merge stable

Signed-off-by: Nicholas Wilson <thewilsonator@users.noreply.github.com>
Merged-on-behalf-of: Iain Buclaw <ibuclaw@users.noreply.github.com>

dkorpel committed:
Deduplicate return parameter destination logic (#14885)

dkorpel committed:
Move incorrectly places runnable tests

ntrel committed:
Fix Issue 23669 - [DIP1000] Compound assignment to length of slice me… (#14865)

Fix Issue 23669 - [DIP1000] Compound assignment to length of slice me…

Signed-off-by: Dennis <dkorpel@users.noreply.github.com>
Merged-on-behalf-of: Dennis <dkorpel@users.noreply.github.com>

ntrel committed:
Tweak semicolon expected to close `alias` declaration error

Show unexpected token.

WalterBright committed:
fix Issue 23711 - compilable/testcstuff1.c:63:1: error: invalid use of restrict

WalterBright committed:
fix Issue 23717 - runnable/bitfields.c:192:5: error: unknown type name S; use struct keyword to refer to the typ

dkorpel committed:
Fix 22916 - copy of ref return still treated as scope variable (#14871)

WalterBright committed:
ImportC: smoke test all C Standard .h files (#14810)

ibuclaw committed:
Merge remote-tracking branch 'upstream/stable' into merge_stable

dlang-bot committed:
Merge pull request #14904 from ibuclaw/merge_stable

merge stable

Signed-off-by: Nicholas Wilson <thewilsonator@users.noreply.github.com>
Merged-on-behalf-of: Nicholas Wilson <thewilsonator@users.noreply.github.com>

WalterBright committed:
add core.factory (#14888)

dkorpel committed:
Revert "fix Issue 23558 - add __traits(getModuleClasses [, module name]) (#14699)" (#14892)

This reverts commit 450ba046cf573768df40a966819d50b3cb1f3f89.

dkorpel committed:
Deprecate `-preview=dip25` switch (#14891)

dkorpel committed:
Fix 23676 - Static foreach hangs compilation for some time (#14886)
